src/NHibernate/Transform/QueryAliasToObjectPropertySetter.cs

```diff
@@ -0,0 +1,197 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace NHibernate.Transform
+{
+	/// <summary>
+	/// Resolves setter for an alias with a heuristic: search among properties then fields for matching name and case, then,
+	/// if no matching property or field was found, retry with a case insensitive match. For members having the same name, it
+	/// sorts them by inheritance depth then by visibility from public to private, and takes those ranking first.
+	/// </summary>
+	[Serializable]
+	public class QueryAliasToObjectPropertySetter
+	{
+		private readonly IDictionary<string, NamedMember<FieldInfo>> _fieldsByNameCaseSensitive;
+		private readonly IDictionary<string, NamedMember<FieldInfo>> _fieldsByNameCaseInsensitive;
+		private readonly IDictionary<string, NamedMember<PropertyInfo>> _propertiesByNameCaseSensitive;
+		private readonly IDictionary<string, NamedMember<PropertyInfo>> _propertiesByNameCaseInsensitive;
+
+		private QueryAliasToObjectPropertySetter(IDictionary<string, NamedMember<FieldInfo>> fieldsByNameCaseSensitive,
+			IDictionary<string, NamedMember<FieldInfo>> fieldsByNameCaseInsensitive,
+			IDictionary<string, NamedMember<PropertyInfo>> propertiesByNameCaseSensitive,
+			IDictionary<string, NamedMember<PropertyInfo>> propertiesByNameCaseInsensitive)
+		{
+			_fieldsByNameCaseSensitive = fieldsByNameCaseSensitive;
+			_fieldsByNameCaseInsensitive = fieldsByNameCaseInsensitive;
+			_propertiesByNameCaseSensitive = propertiesByNameCaseSensitive;
+			_propertiesByNameCaseInsensitive = propertiesByNameCaseInsensitive;
+		}
+
+		/// <summary>
+		/// Set the value of a property or field matching an alias.
+		/// </summary>
+		/// <param name="alias">The alias for which resolving the property or field.</param>
+		/// <param name="value">The value to which the property or field should be set.</param>
+		/// <param name="resultObj">The object on which to set the property or field. It must be of the type for which
+		/// this instance has been built.</param>
+		/// <exception cref="PropertyNotFoundException">Thrown if no matching property or field can be found.</exception>
+		/// <exception cref="AmbiguousMatchException">Thrown if many matching properties or fields are found, having the
+		/// same visibility and inheritance depth.</exception>
+		public void SetProperty(string alias, object value, object resultObj)
+		{
+			if (_propertiesByNameCaseSensitive.TryGetValue(alias, out var property))
+			{
+				checkMember(property);
```

Comment:
On case sensitive case, we should never have an ambiguous case, but better be on the safe side.